File: src/test/java/ma/kalinin/format/RomanNumeralsConverterTest.java
package ma.kalinin.format;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class RomanNumeralsConverterTest {

    private record ArabicToRoman(Integer arabic, String roman) {}
    private static Stream<ArabicToRoman> toRomanExpectations() {
        return Stream.of(
                new ArabicToRoman(1, "I"),
                new ArabicToRoman(11, "XI"),
                new ArabicToRoman(14, "XIV"),
                new ArabicToRoman(86, "LXXXVI"),
                new ArabicToRoman(1990, "MCMXC"),
                new ArabicToRoman(2008, "MMVIII"),
                new ArabicToRoman(1666, "MDCLXVI"),
                new ArabicToRoman(3999, "MMMCMXCIX")
        );
    }

    private record RomanToArabic(String roman, Integer arabic) {}
    private static Stream<RomanToArabic> toArabicExpectations() {
        return toRomanExpectations().map(e -> new RomanToArabic(e.roman, e.arabic));
    }

    @ParameterizedTest
    @MethodSource("toRomanExpectations")
    void toRoman(ArabicToRoman expectation) {
        String actual = RomanNumeralsConverter.toRoman(expectation.arabic);
        assertEquals(expectation.roman, actual);
    }

    @ParameterizedTest
    @MethodSource("toArabicExpectations")
    void toRoman(RomanToArabic expectation) {
        Integer actual = RomanNumeralsConverter.toArabic(expectation.roman);
        assertEquals(expectation.arabic, actual);
    }
}

File: src/test/kotlin/ma/kalinin/format/RomanNumeralsConverterTest.kt
import ma.kalinin.format.toArabic
import ma.kalinin.format.toRoman
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.MethodSource
import kotlin.test.assertEquals

internal class RomanNumeralsConverterTest {
    data class ArabicToRoman(val arabic: Int, val roman: String)
    data class RomanToArabic(val roman: String, val arabic: Int)
    private companion object {
        val pairs = listOf(
            1 to "I", 11 to "XI", 14 to "XIV", 86 to "LXXXVI", 1990 to "MCMXC", 2008 to "MMVIII", 1666 to "MDCLXVI",
            3999 to "MMMCMXCIX",
        )
        val arabicToRomanExpectations: List<ArabicToRoman> = pairs.map { p -> ArabicToRoman(p.first, p.second) }
        val romanToArabicExpectations: List<RomanToArabic> = pairs.map { p -> RomanToArabic(p.second, p.first) }

        // both functions are for @MethodSource since it can't work on collections directly
        @JvmStatic fun toRoman() = arabicToRomanExpectations
        @JvmStatic fun toArabic() = romanToArabicExpectations
    }

    @ParameterizedTest
    @MethodSource
    fun toRoman(expectation: ArabicToRoman) {
        assertEquals(expectation.roman, toRoman(expectation.arabic))
    }

    @ParameterizedTest
    @MethodSource
    fun toArabic(expectation: RomanToArabic) {
        assertEquals(expectation.arabic, toArabic(expectation.roman))
    }
}